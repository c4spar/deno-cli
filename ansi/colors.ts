import * as stdColors from "https://deno.land/std@0.82.0/fmt/colors.ts";

type ExcludedColorMethods = "setColorEnabled" | "getColorEnabled";
type PropertyNames = keyof typeof stdColors;
type ColorMethod = (str: string, ...args: Array<unknown>) => string;
type ColorMethods = Exclude<PropertyNames, ExcludedColorMethods>;
type Chainable<T, E extends keyof T | null = null> = {
  [P in keyof T]: P extends E ? T[P] : Chainable<T, E> & T[P];
};

/** Chainable colors instance returned by all ansi escape properties. */
export type ColorsChain = Chainable<typeof stdColors, ExcludedColorMethods>;

/** Create new `Colors` instance. */
export type ColorsFactory = () => Colors;

/**
 * Chainable colors module.
 * If invoked as method, a new `Colors` instance will be returned.
 */
export type Colors = ColorsFactory & ColorsChain;

/**
 * Chainable colors module.
 * ```
 * console.log(colors.blue.bgRed.bold('Welcome to Deno.Land!'));
 * ```
 * If invoked as method, a new Ansi instance will be returned.
 * ```
 * const myColors: Colors = colors();
 * console.log(myColors.blue.bgRed.bold('Welcome to Deno.Land!'));
 * ```
 */
export const colors: Colors = factory();

function factory(): Colors {
  let stack: Array<ColorMethods> = [];

  const colors: Colors = function (
    this: ColorsChain | undefined,
    str: string,
    ...args: Array<unknown>
  ): string | ColorsChain {
    if (this) {
      const lastname = stack.pop() as ColorMethods;
      const method = stdColors[lastname] as ColorMethod;
      const result: string = method(
        stack.reduce(
          (str: string, name) => (stdColors[name] as ColorMethod)(str),
          str,
        ),
        ...args,
      );
      stack = [];
      return result;
    }
    return factory();
  } as Colors;

  const methodNames = Object.keys(stdColors) as Array<PropertyNames>;

  for (const name of methodNames) {
    if (name === "setColorEnabled" || name === "getColorEnabled") {
      continue;
    }
    Object.defineProperty(colors, name, {
      get(this: ColorsChain) {
        stack.push(name);
        return this;
      },
    });
  }

  return colors;
}
