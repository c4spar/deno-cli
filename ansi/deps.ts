export { encodeBase64 } from "https://deno.land/std@0.216.0/encoding/base64.ts";
export type {
  ReaderSync,
  WriterSync,
} from "https://deno.land/std@0.216.0/io/types.ts";
