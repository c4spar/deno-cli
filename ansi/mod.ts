export * from "./ansi.ts";
export * from "./ansi_escapes.ts";
export * from "./tty.ts";
