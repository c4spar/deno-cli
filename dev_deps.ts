/* std */
export {
  assert,
  assertEquals,
  assertInstanceOf,
  assertRejects,
  assertStrictEquals,
  assertThrows,
} from "https://deno.land/std@0.167.0/testing/asserts.ts";
export {
  assertSpyCalls,
  spy,
} from "https://deno.land/std@0.167.0/testing/mock.ts";
export { assertSnapshot } from "https://deno.land/std@0.167.0/testing/snapshot.ts";
export { describe, it } from "https://deno.land/std@0.167.0/testing/bdd.ts";
export {
  bold,
  red,
  stripColor,
} from "https://deno.land/std@0.167.0/fmt/colors.ts";
export { dirname } from "https://deno.land/std@0.167.0/path/mod.ts";
export { expandGlob } from "https://deno.land/std@0.167.0/fs/mod.ts";
export type { WalkEntry } from "https://deno.land/std@0.167.0/fs/mod.ts";
export { copy } from "https://deno.land/std@0.167.0/streams/conversion.ts";

/* 3rd party */
export { gt, lt } from "https://deno.land/x/semver@v1.4.1/mod.ts";
export { default as sinon } from "https://cdn.skypack.dev/sinon@v13.0.2?dts";
