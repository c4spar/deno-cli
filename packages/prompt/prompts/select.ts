import { blue, dim } from 'https://deno.land/std@v0.41.0/fmt/colors.ts';
import { KeyEvent } from '../../keycode/lib/key-event.ts';
import { Figures } from '../lib/figures.ts';
import { GenericList, GenericListItemOptions, GenericListItemSettings, GenericListNamedItemOptions, GenericListPromptOptions, GenericListPromptSettings } from '../lib/generic-list.ts';

export interface SelectItemOptions extends GenericListItemOptions {
}

export interface SelectNamedItemOptions extends GenericListNamedItemOptions {
}

export interface SelectItemSettings extends GenericListItemSettings {
}

export type SelectValueOptions = ( string | SelectNamedItemOptions )[] | { [ s: string ]: string | SelectItemOptions };
export type SelectValueSettings = SelectItemSettings[];

export interface SelectPromptOptions extends GenericListPromptOptions<string, string> {
    indent?: string;
    pointer?: string;
}

export interface SelectPromptSettings extends GenericListPromptSettings<string, string> {
    indent: string;
    pointer: string;
}

export class Select<S extends SelectPromptSettings> extends GenericList<string, string, S> {

    protected selected: number = typeof this.settings.default !== 'undefined' ? this.settings.values.findIndex( item => item.name === this.settings.default ) || 0 : 0;

    public static async prompt( options: SelectPromptOptions ): Promise<string | undefined> {

        const items: SelectNamedItemOptions[] = this.mapValues( options.values );
        const values: SelectValueSettings = items.map( item => this.mapItem( item ) );

        return new this( {
            pointer: blue( Figures.POINTER ),
            indent: ' ',
            maxRows: 10,
            ...options,
            values
        } ).run();
    }

    public static separator( label: string = '------------' ): SelectNamedItemOptions {
        return super.separator();
    }

    protected static mapValues( optValues: SelectValueOptions ): SelectNamedItemOptions[] {
        return super.mapValues( optValues ) as SelectNamedItemOptions[];
    }

    protected static mapItem( item: SelectNamedItemOptions ): SelectItemSettings {
        return super.mapItem( item ) as SelectItemSettings;
    }

    protected getMessage(): string {

        let message = this.settings.message;

        if ( typeof this.settings.default !== 'undefined' ) {
            message += dim( ` (${ this.settings.default })` );
        }

        return message;
    }

    protected async handleEvent( event: KeyEvent ): Promise<boolean> {

        switch ( event.name ) {

            case 'c':
                if ( event.ctrl ) {
                    return Deno.exit( 0 );
                }
                break;

            case 'up':
                await this.selectPrevious();
                break;

            case 'down':
                await this.selectNext();
                break;

            case 'return':
            case 'enter':
                return this.selectValue();
        }

        return false;
    }

    protected async selectValue() {
        return this.validateValue( this.settings.values[ this.selected ].name );
    }

    protected writeListItem( item: SelectItemSettings, isSelected?: boolean ) {

        let line = this.settings.indent;

        // pointer
        line += isSelected ? `${ this.settings.pointer } ` : '  ';

        // value
        const value: string = item.label;
        line += `${ isSelected ? value : dim( value ) }`;

        this.writeLine( line );
    }

    protected sanitize( value: string ): string {
        return value;
    }

    protected validate( value: string ): boolean {
        return true;
    }

    protected transform( value: string ): string {
        return value;
    }
}
