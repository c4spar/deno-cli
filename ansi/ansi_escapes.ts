import { encodeBase64 } from "./deps.ts";

/** Image options. */
export interface ImageOptions {
  width?: number;
  height?: number;
  preserveAspectRatio?: boolean;
}

/** Escape sequence: `\x1B` */
const ESC = "\x1B";
/** Control sequence intro: `\x1B[` */
const CSI = `${ESC}[`;
/** Operating system command: `\x1B]` */
const OSC = `${ESC}]`;
/** Link separator */
const SEP = ";";

export const ansiEscapes = {
  /** Ring audio bell: `\u0007` */
  bel: "\u0007",
  /** Get cursor position. */
  cursorPosition: `${CSI}6n`,
  /**
   * Move cursor to x, y, counting from the top left corner.
   * @param x Position left.
   * @param y Position top.
   */
  cursorTo(x: number, y?: number): string {
    if (typeof y !== "number") {
      return `${CSI}${x}G`;
    }
    return `${CSI}${y};${x}H`;
  },
  /**
   * Move cursor by offset.
   * @param x Offset left.
   * @param y Offset top.
   */
  cursorMove(x: number, y: number): string {
    let ret = "";

    if (x < 0) {
      ret += `${CSI}${-x}D`;
    } else if (x > 0) {
      ret += `${CSI}${x}C`;
    }

    if (y < 0) {
      ret += `${CSI}${-y}A`;
    } else if (y > 0) {
      ret += `${CSI}${y}B`;
    }

    return ret;
  },
  /**
   * Move cursor up by n lines.
   * @param count Number of lines.
   */
  cursorUp: (count = 1): string => `${CSI}${count}A`,
  /**
   * Move cursor down by n lines.
   * @param count Number of lines.
   */
  cursorDown: (count = 1): string => `${CSI}${count}B`,
  /**
   * Move cursor forward by n lines.
   * @param count Number of lines.
   */
  cursorForward: (count = 1): string => `${CSI}${count}C`,
  /**
   * Move cursor backward by n lines.
   * @param count Number of lines.
   */
  cursorBackward: (count = 1): string => `${CSI}${count}D`,
  /**
   * Move cursor to the beginning of the line n lines down.
   * @param count Number of lines.
   */
  cursorNextLine: (count = 1): string => `${CSI}E`.repeat(count),
  /**
   * Move cursor to the beginning of the line n lines up.
   * @param count Number of lines.
   */
  cursorPrevLine: (count = 1): string => `${CSI}F`.repeat(count),
  /** Move cursor to first column of current row. */
  cursorLeft: `${CSI}G`,
  /** Hide cursor. */
  cursorHide: `${CSI}?25l`,
  /** Show cursor. */
  cursorShow: `${CSI}?25h`,
  /** Save cursor. */
  cursorSave: `${ESC}7`,
  /** Restore cursor. */
  cursorRestore: `${ESC}8`,
  /**
   * Scroll window up by n lines.
   * @param count Number of lines.
   */
  scrollUp: (count = 1): string => `${CSI}S`.repeat(count),
  /**
   * Scroll window down by n lines.
   * @param count Number of lines.
   */
  scrollDown: (count = 1): string => `${CSI}T`.repeat(count),
  /** Clear screen. */
  eraseScreen: `${CSI}2J`,
  /**
   * Clear screen up.
   * @param count Number of lines.
   */
  eraseUp: (count = 1): string => `${CSI}1J`.repeat(count),
  /**
   * Clear screen down.
   * @param count Number of lines.
   */
  eraseDown: (count = 1): string => `${CSI}0J`.repeat(count),
  /** Clear current line. */
  eraseLine: `${CSI}2K`,
  /** Clear to line end. */
  eraseLineEnd: `${CSI}0K`,
  /** Clear to line start. */
  eraseLineStart: `${CSI}1K`,
  /**
   * Clear screen and move cursor by n lines up and move cursor to first column.
   * @param count Number of lines.
   */
  eraseLines(count: number): string {
    let clear = "";
    for (let i = 0; i < count; i++) {
      clear += ansiEscapes.eraseLine +
        (i < count - 1 ? ansiEscapes.cursorUp() : "");
    }
    clear += ansiEscapes.cursorLeft;
    return clear;
  },
  /**
   * Create link.
   * @param text Link text.
   * @param url Link url.
   * ```
   * console.log(
   *   ansi.link("Click me.", "https://deno.land"),
   * );
   * ```
   */
  link: (text: string, url: string): string =>
    [
      OSC,
      "8",
      SEP,
      SEP,
      url,
      ansiEscapes.bel,
      text,
      OSC,
      "8",
      SEP,
      SEP,
      ansiEscapes.bel,
    ].join(""),
  /**
   * Create image.
   * @param buffer  Image buffer.
   * @param options Image options.
   * ```
   * const response = await fetch("https://deno.land/images/hashrock_simple.png");
   * const imageBuffer: ArrayBuffer = await response.arrayBuffer();
   * console.log(
   *   ansi.image(imageBuffer),
   * );
   * ```
   */
  image(
    buffer: string | ArrayBuffer,
    options?: ImageOptions,
  ): string {
    let ret = `${OSC}1337;File=inline=1`;

    if (options?.width) {
      ret += `;width=${options.width}`;
    }

    if (options?.height) {
      ret += `;height=${options.height}`;
    }

    if (options?.preserveAspectRatio === false) {
      ret += ";preserveAspectRatio=0";
    }

    return ret + ":" + encodeBase64(buffer) + ansiEscapes.bel;
  },
};
