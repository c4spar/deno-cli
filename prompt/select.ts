import { brightBlue, underline } from "./deps.ts";
import {
  GenericList,
  GenericListKeys,
  GenericListOption,
  GenericListOptionGroup,
  GenericListOptionGroupSettings,
  GenericListOptions,
  GenericListOptionSettings,
  GenericListSettings,
  isOption,
  isOptionGroup,
} from "./_generic_list.ts";
import { GenericPrompt } from "./_generic_prompt.ts";
import { getFiguresByKeys } from "./figures.ts";

/** Select prompt options. */
export interface SelectOptions extends GenericListOptions<string, string> {
  options: Array<string | SelectOption | SelectOptionGroup>;
  keys?: SelectKeys;
}

/** Select prompt settings. */
export interface SelectSettings extends
  GenericListSettings<
    string,
    string,
    SelectOptionSettings,
    SelectOptionGroupSettings
  > {
  keys: SelectKeys;
}

/** Select option options. */
export type SelectOption = GenericListOption;

/** Select option group options. */
export type SelectOptionGroup = GenericListOptionGroup<GenericListOption>;

/** Select option settings. */
export type SelectOptionSettings = GenericListOptionSettings;

/** Select option group settings. */
export type SelectOptionGroupSettings = GenericListOptionGroupSettings<
  SelectOptionSettings
>;

/** Select key options. */
export type SelectKeys = GenericListKeys;

/** Select prompt representation. */
export class Select extends GenericList<
  string,
  string,
  SelectOptionSettings,
  SelectOptionGroupSettings
> {
  protected readonly settings: SelectSettings;
  protected options: Array<SelectOptionSettings | SelectOptionGroupSettings>;
  protected listIndex: number;
  protected listOffset: number;

  /** Execute the prompt and show cursor on end. */
  public static prompt(options: SelectOptions): Promise<string> {
    return new this(options).prompt();
  }

  /**
   * Inject prompt value. Can be used for unit tests or pre selections.
   * @param value Input value.
   */
  public static inject(value: string): void {
    GenericPrompt.inject(value);
  }

  constructor(options: SelectOptions) {
    super();
    this.settings = this.getDefaultSettings(options);
    this.options = this.settings.options.slice();
    this.listIndex = this.getListIndex(this.settings.default);
    this.listOffset = this.getPageOffset(this.listIndex);
  }

  protected getDefaultSettings(options: SelectOptions): SelectSettings {
    return {
      ...super.getDefaultSettings(options),
      options: this.mapOptions(options, options.options),
    };
  }

  /**
   * Map string option values to options and set option defaults.
   * @param promptOptions Select options.
   */
  protected mapOptions(
    promptOptions: SelectOptions,
    options: Array<string | SelectOption | SelectOptionGroup>,
  ): Array<SelectOptionSettings | SelectOptionGroupSettings> {
    return options.map((option) =>
      isOptionGroup(option)
        ? this.mapOptionGroup(promptOptions, option)
        : typeof option === "string"
        ? this.mapOption(promptOptions, { value: option })
        : this.mapOption(promptOptions, option)
    );
  }

  protected input(): string {
    return underline(brightBlue(this.inputValue));
  }

  protected async submit(): Promise<void> {
    if (
      this.isBackButton(this.selectedOption) ||
      isOptionGroup(this.selectedOption)
    ) {
      const info = isOptionGroup(this.selectedOption)
        ? ` To select a group use ${
          getFiguresByKeys(this.settings.keys.open ?? []).join(", ")
        }.`
        : "";
      this.setErrorMessage(`No option selected.${info}`);
      return;
    }

    await super.submit();
  }

  /** Get value of selected option. */
  protected getValue(): string {
    const option = this.options[this.listIndex];
    assertIsOption(option);
    return option?.value ?? this.settings.default;
  }

  /**
   * Validate input value.
   * @param value User input value.
   * @return True on success, false or error message on error.
   */
  protected validate(value: string): boolean | string {
    return typeof value === "string" &&
      value.length > 0 &&
      this.options.findIndex((
          option: SelectOptionSettings | SelectOptionGroupSettings,
        ) => isOption(option) && option.value === value
        ) !== -1;
  }

  /**
   * Map input value to output value.
   * @param value Input value.
   * @return Output value.
   */
  protected transform(value: string): string {
    return value.trim();
  }

  /**
   * Format output value.
   * @param value Output value.
   */
  protected format(value: string): string {
    return this.getOptionByValue(value)?.name ?? value;
  }
}

function assertIsOption<
  TOption extends GenericListOption,
>(
  option: TOption | GenericListOptionGroup<GenericListOption>,
): asserts option is TOption {
  if (!isOption(option)) {
    throw new Error("Expected an option but got an option group.");
  }
}

/**
 * Select options type.
 * @deprecated Use `Array<string | SelectOption | SelectOptionGroup>` instead.
 */
export type SelectValueOptions = Array<
  string | SelectOption | SelectOptionGroup
>;

/**
 * Select option settings type.
 * @deprecated Use `Array<SelectOptionSettings | SelectOptionGroupSettings>` instead.
 */
export type SelectValueSettings = Array<
  SelectOptionSettings | SelectOptionGroupSettings
>;
