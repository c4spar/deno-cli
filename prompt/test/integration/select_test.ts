import { ansi } from "../../../ansi/ansi.ts";
import { format } from "../../../dev_deps.ts";
import { Select } from "../../select.ts";
import { snapshotTest } from "../../../testing/snapshot.ts";

await snapshotTest({
  name: "select prompt > should select an option",
  meta: import.meta,
  osSuffix: ["windows"],
  stdin: ansi
    .cursorDown
    .cursorDown
    .text("\n")
    .toArray(),
  async fn() {
    await Select.prompt({
      message: "Select an option",
      options: [
        { name: "Foo", value: "foo" },
        { name: "Bar", value: "bar" },
        { name: "Baz", value: "baz" },
      ],
    });
  },
});

await snapshotTest({
  name: "select prompt > should search an option",
  meta: import.meta,
  osSuffix: ["windows"],
  stdin: ansi
    .text("baz")
    .text("\n")
    .toArray(),
  async fn() {
    await Select.prompt({
      message: "Select an option",
      search: true,
      options: [
        { name: "Foo", value: "foo" },
        { name: "Bar", value: "bar" },
        { name: "Baz", value: "baz" },
      ],
    });
  },
});

await snapshotTest({
  name: "select prompt > should format option value",
  meta: import.meta,
  osSuffix: ["windows"],
  stdin: ansi
    .cursorDown
    .text("\n")
    .toArray(),
  async fn() {
    await Select.prompt({
      message: "Message...",
      options: [
        { value: new Date(10000) },
        { value: new Date(20000) },
      ],
      format: (date) => format(date, "dd-MM-yyyy"),
    });
  },
});

await snapshotTest({
  name: "select prompt > should support separator option",
  meta: import.meta,
  osSuffix: ["windows"],
  stdin: ansi
    .cursorDown
    .text("\n")
    .toArray(),
  async fn() {
    await Select.prompt({
      message: "Message...",
      options: [
        { value: new Date(10000) },
        Select.separator("+++++"),
        { value: new Date(20000) },
      ],
      format: (date) => format(date, "dd-MM-yyyy"),
    });
  },
});
