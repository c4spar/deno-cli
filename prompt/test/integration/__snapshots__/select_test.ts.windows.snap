export const snapshot = {};

snapshot[`select prompt > should select an option 1`] = `
stdout:
"? Select an option
❯ Foo
  Bar
  Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  Foo
❯ Bar
  Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option
  Foo
  Bar
❯ Baz\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Select an option » Baz
\\x1b[?25h\\x1b[?25h"
stderr:
""
`;

snapshot[`select prompt > should search an option 1`] = `
stdout:
"? Select an option 🔎 
❯ Foo
  Bar
  Baz\\x1b[3A\\x1b[23G\\x1b[G\\x1b[0J
? Select an option 🔎 baz
❯ Baz\\x1b[1A\\x1b[26G\\x1b[G\\x1b[0J
? Select an option » Baz
\\x1b[?25h\\x1b[?25h"
stderr:
""
`;

snapshot[`select prompt > should format option value 1`] = `
stdout:
"? Message...
❯ 01-01-1970
  01-01-1970\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Message...
  01-01-1970
❯ 01-01-1970\\x1b[2A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Message... » 01-01-1970
\\x1b[?25h\\x1b[?25h"
stderr:
""
`;

snapshot[`select prompt > should support separator option 1`] = `
stdout:
"? Message...
❯ 01-01-1970
  +++++
  01-01-1970\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Message...
  01-01-1970
  +++++
❯ 01-01-1970\\x1b[3A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Message... » 01-01-1970
\\x1b[?25h\\x1b[?25h"
stderr:
""
`;
